### ISSUE TITLE
How lightgbm calculate gain and feature importance when objective is quantile or mean absolute error?

### ISSUE THREAD

**Issue by longyunshen** [microsoft/LightGBM#2951]
0

I know lightgbm is kind of second order taylor expansion to boost trees targetting to reduce loss function. I am trying to figure how lightgbm deals with quantile regression when calculate gains.

When objective function is normal ols, in sklearn feature importance is calcuated by negative gradient which is order of original y. You can check the source code, even sklearn's feature importance only return relative values which are summed to 1. In lightgbm, feature importance is calcuated by raw (y-y_predicted)^2. It means that if I increase y ten times larger, the gain will be 100 times larger. And I validate it with following code. It is ok with me.

When objective comming to quantile or mean absolute error, I get lost. In sklearn, only negative gradient is considered, which means the magnitude of y is ignored. And the gain is numerically very small, as only quantile is considered. It is around magnitude of 0.000 something. In lightgbm, it is of magnitue of 90 or something. And the gain or feature importance does not change when i change the magnitude of y. Can someone elaberate me How lightgbm calculate gain and feature importance when objective is quantile or mean absolute error? Thank you very much . The following code is for test.

There is a link to describe it. But I still dont get it. http://jmarkhou.com/lgbqr/
The following code is to reprocduce the gain of tree. The gain stay unchanged when I change the magnitude of target. 

import pandas as pd
import matplotlib.pyplot as plt
import numpy as np
import lightgbm as lgb

from sklearn.metrics import r2_score, mean_absolute_error

np.random.seed(0)

n=100
x1=np.random.normal(0,10,n)
x2=np.random.normal(0,10,n)

X = np.array([x1,x2]).T
y = x1 +x2
y2 = 10*y
'''
params = {'objective':'regression',
          'num_leaves':2,
          'n_estimators':2
          }
'''

params = {'objective':'l1',
          'num_leaves':2,
          'n_estimators':2
          }

lgb_train=lgb.Dataset(X,y)
lgb_train2=lgb.Dataset(X,y2)

gbm = lgb.train(params,lgb_train)

gbm2 = lgb.train(params,lgb_train2)

predict1 = gbm.predict(X)
predict2 = gbm2.predict(X)

print('Plotting model1 tree1...')  # one tree use categorical feature to split
ax = lgb.plot_tree(gbm, tree_index=0, figsize=(10, 10), show_info=['split_gain'])
plt.show()

print('Plotting model1 tree1...')  # one tree use categorical feature to split
ax = lgb.plot_tree(gbm, tree_index=1, figsize=(10, 10), show_info=['split_gain'])
plt.show()

print('Plotting model2 tree2...')  # one tree use categorical feature to split
ax = lgb.plot_tree(gbm2, tree_index=0, figsize=(10, 10), show_info=['split_gain'])
plt.show()

print('Plotting  tree...')  # one tree use categorical feature to split
ax = lgb.plot_tree(gbm2, tree_index=1, figsize=(10, 10), show_info=['split_gain'])
plt.show()

print('mean_absolute_error model1',mean_absolute_error(y,predict1))
print('mean_absolute_error model2',mean_absolute_error(y2,predict2))

**Comment by guolinke** [comment]
in mae or quantlie, the first order gradient is constant (-1 or 1), whatever the y is, and the second order gradient is always `1`.
Therefore, the gain is about `sign(y - y_pred)`.

**Comment by longyunshen** [comment]
> in mae or quantlie, the first order gradient is constant (-1 or 1), whatever the y is, and the second order gradient is always `1`.
> Therefore, the gain is about `sign(y - y_pred)`.

Thanks for the quick reply. Why use sign(y - y_pred) as gain, rather than original loss function sum(abs(y-y_pred))?  The difference can mean something?

**Comment by longyunshen** [comment]
> in mae or quantlie, the first order gradient is constant (-1 or 1), whatever the y is, and the second order gradient is always `1`.
> Therefore, the gain is about `sign(y - y_pred)`.

why sign(y-y_pred) can have gain as float? if you run the code, you will find the gain of first tree is 50.776. Thanks for your reply!

**Comment by guolinke** [comment]
I said it is "about" (related to) `sign(y - y_pred)`, not exactly it is.
The real gain is https://github.com/microsoft/LightGBM/blob/c870c57a14b61329bf493b3b1a9a2893a62405cf/src/treelearner/feature_histogram.hpp#L714-L746
